### DISCUSSION TITLE
Why no macros to ensure request handlers are correct?

### DISCUSSION THREAD

**Discussion by darkwater** [tokio-rs/axum#2151] [Q&A]
I was wondering why the decision was made to not check path parameters of request handlers at compile-time. Alternatives such as Rocket use a proc macro on request handlers and can ensure the function signature catches the path parameters, eg.:

```rust
#[get("/foo/<bar>")]
async fn get_handler(bar: u32) -> ...
```

While axum involves no macros here and is unable to check this, and doesn't even check all routes when the server starts; it waits until a handler is actually invoked.

I understand avoiding macros when not necessary, but this seems like a decision that isn't really in the spirit of Rust. Is there a particular reason behind this decision?

**Comment by davidpdrsn** [comment]
If you search around you can find previous answers to this question. Mostly that it makes it harder to compose stuff. You can only abstract over macros with more macros.

**Reply by davidpdrsn** [reply]
We also have https://docs.rs/axum-extra/0.7.7/axum_extra/routing/trait.TypedPath.html which is a better solution than rocket imo. Mostly from being a derive macro and not an attribute.
